src/System.Net.Security/src/System/Security/Authentication/ExtendedProtection/ServiceNameCollection.cs

```diff
@@ -0,0 +1,205 @@
+        {
+            Debug.Assert(serviceNames != null);
+            Debug.Assert(!String.IsNullOrEmpty(searchServiceName));
+
+            foreach (string serviceName in serviceNames)
+            {
+                if (Match(serviceName, searchServiceName))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public bool Contains(string searchServiceName)
+        {
+            string searchName = NormalizeServiceName(searchServiceName);
+
+            return Contains(searchName, InnerList);
+        }
+
+        // Normalizes any punycode to unicode in an Service Name (SPN) host.
+        // If the algorithm fails at any point then the original input is returned.
+        // ServiceName is in one of the following forms:
+        // prefix/host
+        // prefix/host:port
+        // prefix/host/DistinguishedName
+        // prefix/host:port/DistinguishedName
+        internal static string NormalizeServiceName(string inputServiceName)
+        {
+            if (string.IsNullOrWhiteSpace(inputServiceName))
+            {
+                return inputServiceName;
+            }
+
+            // Separate out the prefix
+            int shashIndex = inputServiceName.IndexOf('/');
+            if (shashIndex < 0)
+            {
+                return inputServiceName;
+            }
+            string prefix = inputServiceName.Substring(0, shashIndex + 1); // Includes slash
+            string hostPortAndDistinguisher = inputServiceName.Substring(shashIndex + 1); // Excludes slash
+
+            if (string.IsNullOrWhiteSpace(hostPortAndDistinguisher))
+            {
+                return inputServiceName;
+            }
+
+            string host = hostPortAndDistinguisher;
+            string port = string.Empty;
+            string distinguisher = string.Empty;
+
+            // Check for the absence of a port or distinguisher.
+            UriHostNameType hostType = Uri.CheckHostName(hostPortAndDistinguisher);
+            if (hostType == UriHostNameType.Unknown)
+            {
+                string hostAndPort = hostPortAndDistinguisher;
+
+                // Check for distinguisher.
+                int nextSlashIndex = hostPortAndDistinguisher.IndexOf('/');
+                if (nextSlashIndex >= 0)
+                {
+                    // host:port/distinguisher or host/distinguisher
+                    hostAndPort = hostPortAndDistinguisher.Substring(0, nextSlashIndex); // Excludes Slash
+                    distinguisher = hostPortAndDistinguisher.Substring(nextSlashIndex); // Includes Slash
+                    host = hostAndPort; // We don't know if there is a port yet.
+                    // No need to validate the distinguisher.
+                }
+
+                // Check for port.
+                int colonIndex = hostAndPort.LastIndexOf(':'); // Allow IPv6 addresses.
```

Comment:
It's not a TODO, but an explanation on why that line exists.
This is not a URI, but an [SPN](https://msdn.microsoft.com/en-us/library/ms677601(v=vs.85).aspx) so I can't be 100% sure.